### PR TITLE
Nnunet amp integration

### PR DESCRIPTION
**Pull request by jewelltaylor**
# PR Type
[**Feature** | Fix | Documentation | Other ]

# Short Description

Clickup Ticket(s): [Add Mixed Precision Training to NnuentClient](https://app.clickup.com/t/868beh5kf)

By default, if you have a CUDA GPU, nnUNetTrainer uses automatic mixed precision training. This casts down the weights and inputs/activation from fp32 to fp16 for certain sets of operations. Weights are kept in fp32 and downcasted. Gradients are compute in fp16 but accumulated in fp32. In order to realize automatic mixed precision training, we must wrap forward pass and loss computation in the autocast context. Additionally, the loss must be scaled to prevent underflow of the gradients which are computed in fp16. This is accomplished via the torch provided GradScaler.

### CODE REVIEW

File: fl4health/clients/nnunet_client.py [hunk 1]
```diff
@@ -188,6 +189,9 @@ def __init__(
         # Used to redirect stdout to logger
         self.stream2debug = StreamToLogger(FLOWER_LOGGER, DEBUG)
 
+        # Used to scale gradients if using mized precision training (true if device is cuda)
```
Comment (scarere) [inline fl4health/clients/nnunet_client.py:192]: nit: mixed

File: fl4health/clients/nnunet_client.py [hunk 2]
```diff
@@ -203,6 +207,58 @@ def __init__(
                 log(INFO, "Switching pytorch model jit compile to OFF")
             os.environ["nnUNet_compile"] = str("false")
 
+    def train_step(self, input: TorchInputType, target: TorchTargetType) -> Tuple[TrainingLosses, TorchPredType]:
```
Comment (scarere) [inline fl4health/clients/nnunet_client.py:210]: I'm not super familiar with mixed precision training so I'm going to trust that this does what its supposed to do.

File: fl4health/clients/nnunet_client.py [hunk 3]
```diff
@@ -15,6 +15,7 @@
 from flwr.common.logger import FLOWER_LOGGER, console_handler, log
 from flwr.common.typing import Config, Scalar
 from torch import nn
+from torch.cuda.amp import GradScaler
```
Comment (scarere) [inline fl4health/clients/nnunet_client.py:18]: Curious on your thoughts on whether mixed precision training should be specific to nnunet, or whether its a feature that should be extended to Basic Client (or maybe it warrants a new client subclass and the nnunet client either subclasses or mixes in the new subclass).

File: fl4health/clients/nnunet_client.py (hunk 1 above)
Comment (scarere) [reply fl4health/clients/nnunet_client.py:192]: Should we add a specific argument here rather than making mixed precision the default for cuda. Is there ever any reason not to use mixed precision?

File: fl4health/clients/nnunet_client.py [hunk 4]
```diff
@@ -525,7 +582,14 @@ def predict(self, input: TorchInputType) -> Tuple[TorchPredType, Dict[str, torch
             unused by this subclass and therefore is always an empty dict
         """
         if isinstance(input, torch.Tensor):
-            output = self.model(input)
+            # If device type is cuda, nnUNet defaults to mixed precision forward pass
+            # https://github.com/MIC-DKFZ/nnUNet/blob/43349fa5f0680e8109a78dca7215c19e258c9dd7 \
+            # nnunetv2/training/nnUNetTrainer/nnUNetTrainer.py#L993
+            if self.device.type == "cuda":
+                with torch.autocast(self.device.type, enabled=True):
```
Comment (scarere) [inline fl4health/clients/nnunet_client.py:589]: This is cool, never seen it used before. I thought mixed precision training would be more involved

File: fl4health/clients/nnunet_client.py [hunk 5]
```diff
@@ -577,7 +642,16 @@ def compute_loss_and_additional_losses(
                 f"Got {len(loss_preds)} predictions and {len(loss_targets)} targets."
             )
 
-        return self.criterion(loss_preds, loss_targets), None
+        # If device type is cuda, nnUNet defaults to compute loss in mixed precision
+        # https://github.com/MIC-DKFZ/nnUNet/blob/43349fa5f0680e8109a78dca7215c19e258c9dd7 \
+        # nnunetv2/training/nnUNetTrainer/nnUNetTrainer.py#L993
+        if self.device.type == "cuda":
+            with torch.autocast(self.device.type, enabled=True):
+                loss = self.criterion(loss_preds, loss_targets), None
```
Comment (scarere) [inline fl4health/clients/nnunet_client.py:650]: Out of curiosity, does autocast automatically cast variables back to the default precision when the scope is exited? For example are the loss values here going to remain fp16 or whatever they were cast to?

**Review by scarere — APPROVED** [review]
PR looks good to me. Most of my comments are just me asking questions about mixed precision training out of interest

File: fl4health/clients/nnunet_client.py (hunk 5 above)
Comment (jewelltaylor) [reply fl4health/clients/nnunet_client.py:650]: Yeah I think you are right!

File: fl4health/clients/nnunet_client.py (hunk 3 above)
Comment (jewelltaylor) [reply fl4health/clients/nnunet_client.py:18]: I will add it in the backlog as a potential addition that the user can optionally specify
